tests/unit/plugins/modules/test_keycloak_userprofile.py

```diff
@@ -0,0 +1,998 @@
+                                        "edit": [
+                                            "admin",
+                                            "user"
+                                        ]
+                                    },
+                                    "multivalued": False,
+                                    "annotations": {}
+                                },
+                                {
+                                    "name": "foobar2",
+                                    "displayName": "foobar2",
+                                    "validations": {
+                                        "length": {
+                                            "max": 255
+                                        },
+                                        "person-name-prohibited-characters": {}
+                                    },
+                                    "required": {
+                                        "roles": [
+                                            "user"
+                                        ]
+                                    },
+                                    "permissions": {
+                                        "view": [
+                                            "admin",
+                                            "user"
+                                        ],
+                                        "edit": [
+                                            "admin",
+                                            "user"
+                                        ]
+                                    },
+                                    "multivalued": False,
+                                    "annotations": {}
+                                }
+                            ],
+                            "groups": [
+                                {
+                                    "name": "user-metadata",
+                                    "displayHeader": "User metadata",
+                                    "displayDescription": "Attributes, which refer to user metadata",
+                                }
+                            ],
+                        }
+                    ]
+                }
+            }
+        ]
+        return_value_component_update = [
+            None
+        ]
+        changed = True
+
+        set_module_args(module_args)
+
+        # Run the module
+
+        with mock_good_connection():
+            with patch_keycloak_api(get_components=return_value_get_components_get, get_component=return_value_component_get,
+                                    update_component=return_value_component_update) as (mock_get_components, mock_get_component, mock_create_component, mock_update_component, mock_delete_component):
+                with self.assertRaises(AnsibleExitJson) as exec_info:
+                    self.module.main()
+
+        self.assertEqual(len(mock_get_components.mock_calls), 2)
```

Comment:
`ks.get_components` is only ever called once in the module (line 510, outside of a loop). Why do you expect it to be called twice?

Reply:
@felixfontein Thanks for the hint! After looking at it again I could make sense of the unit test ;)
Also I fixed all of the remaining sanity checks and reworked the integration tests.

I think this is good for a review.